pkg/alertmanager/amcfg.go

```diff
@@ -161,37 +161,50 @@ func (cg *configGenerator) generateConfig(
 	return yaml.Marshal(generatedConf)
 }
 
-func convertRoute(in *monitoringv1alpha1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+func (cg *configGenerator) convertRoute(in *monitoringv1alpha1.Route, crKey types.NamespacedName, firstLevelRoute bool) *route {
+	var matchers []string
+	matchersV2Allowed := cg.amVersion.GTE(semver.MustParse("0.22.0"))
+
+	v2NamespaceMatcher := monitoringv1alpha1.Matcher{
+		Name:      "namespace",
+		Value:     crKey.Namespace,
+		MatchType: monitoringv1alpha1.MatchEqual,
+	}.String()
+
 	// Enforce "continue" to be true for the top-level route.
 	cont := in.Continue
 	if firstLevelRoute {
 		cont = true
 	}
 
+	// deprecated
 	match := map[string]string{}
 	matchRE := map[string]string{}
 
 	for _, matcher := range in.Matchers {
+		// prefer matchers to deprecated config
+		if matcher.MatchType != "" {
+			matchers = append(matchers, matcher.String())
+			continue
+		}
+
 		if matcher.Regex {
 			matchRE[matcher.Name] = matcher.Value
 		} else {
 			match[matcher.Name] = matcher.Value
 		}
 	}
 	if firstLevelRoute {
-		match["namespace"] = crKey.Namespace
+		if matchersV2Allowed {
+			if !contains(v2NamespaceMatcher, matchers) {
```

Comment:
side-note: we could even not check if the matcher is already present and always append it.